File: src/xqa/commons/xml_file_finder.py
import logging
import mimetypes
import os

from lxml import etree
from lxml.etree import XMLSyntaxError


class XmlFileFinder:
    class FinderException(Exception):

        def __init__(self, message=None):
            if message:
                self.message = message

    ERROR_NO_XML_FILES_FOUND = 'no XML files found'
    ERROR_FILE_MIMETYPE = 'incorrect mimetype'
    ERROR_FILE_CONTENTS_NOT_WELL_FORMED = 'file not well-formed'

    def __init__(self, path_to_xml_candidate_files):
        self._path_to_xml_candidate_files = path_to_xml_candidate_files

    def find_files(self):
        xml_files = []

        for root, _, filenames in os.walk(self._path_to_xml_candidate_files):
            for filename in filenames:
                path_to_filename = self._full_path_to_file(root, filename)
                try:
                    if self.can_file_be_used(path_to_filename, filename):
                        xml_files.append(path_to_filename)
                except XmlFileFinder.FinderException:
                    pass

        if not xml_files:
            logging.warning(XmlFileFinder.ERROR_NO_XML_FILES_FOUND)
            raise XmlFileFinder.FinderException(XmlFileFinder.ERROR_NO_XML_FILES_FOUND)

        return sorted(xml_files)

    def _rm_bom_from_file_contents(self, path_to_filename):
        file_contents = open(path_to_filename, mode='r', encoding='utf-8-sig').read()
        open(path_to_filename, mode='w', encoding='utf-8').write(file_contents)

    def _full_path_to_file(self, root, filename):
        return os.path.join(root, filename)

    def can_file_be_used(self, path_to_filename, filename):
        if not self._check_file_mimetype_recognised(path_to_filename):
            logging.debug('N: %s: %s' % (XmlFileFinder.ERROR_FILE_MIMETYPE, filename))
            raise XmlFileFinder.FinderException(XmlFileFinder.ERROR_FILE_MIMETYPE)

        if not self._check_file_contents_well_formed(path_to_filename):
            logging.debug('N: %s: %s' % (XmlFileFinder.ERROR_FILE_CONTENTS_NOT_WELL_FORMED, filename))
            raise XmlFileFinder.FinderException(XmlFileFinder.ERROR_FILE_CONTENTS_NOT_WELL_FORMED)

        self._rm_bom_from_file_contents(path_to_filename)

        logging.debug('Y: %s' % path_to_filename)
        return True

    def _check_file_contents_well_formed(self, path_to_filename):
        try:
            etree.parse(path_to_filename)
            return True
        except XMLSyntaxError:
            return False

    def _check_file_mimetype_recognised(self, path_to_filename):
        if mimetypes.guess_type(path_to_filename) in [('application/xml', None), ('text/xml', None)]:
            return True
        return False

    def contents_of_file(self, xml_file):
        with open(xml_file) as f:
            return f.read()

File: src/main/java/xqa/ingest/XmlFileFinder.java
package xqa.ingest;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.IOUtils;
import org.apache.commons.io.input.BOMInputStream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.text.MessageFormat;
import java.util.Collection;

public class XmlFileFinder {
    public static final String ERROR_NO_XML_FILES_FOUND = "no XML files found";
    public static final String ERROR_FILE_MIMETYPE = "incorrect mimetype";
    public static final String ERROR_FILE_CONTENTS_NOT_WELL_FORMED = "file not well-formed";

    private static final Logger logger = LoggerFactory.getLogger(XmlFileFinder.class);
    private String pathToXmlCandidateFiles;

    public XmlFileFinder() {
    }

    public XmlFileFinder(final String pathToXmlCandidateFiles) {
        this.pathToXmlCandidateFiles = pathToXmlCandidateFiles;
    }

    public Collection<File> findFiles() throws FinderException {
        Collection<File> candidateXmlFiles = FileUtils.listFiles(new File(pathToXmlCandidateFiles),
                new String[]{"xml", "XML"}, true);

        if (candidateXmlFiles.size() == 0)
            throw new XmlFileFinder.FinderException(XmlFileFinder.ERROR_NO_XML_FILES_FOUND);

        logger.info(
                MessageFormat.format("found {0} file(s) in {1}", candidateXmlFiles.size(), pathToXmlCandidateFiles));

        return candidateXmlFiles;
    }

    public void rmBomFromFileContents(File candidateXmlfile) throws Exception {
        // grep -rl $'\xEF\xBB\xBF' .
        InputStream fileInputStream = new FileInputStream(candidateXmlfile);
        BOMInputStream bomInputStream = new BOMInputStream(fileInputStream);

        File tmpFile = null;
        if (bomInputStream.hasBOM()) {
            logger.debug(MessageFormat.format("removing BOM: {0}", candidateXmlfile.getCanonicalPath()));

            tmpFile = new File(candidateXmlfile.getPath().concat(".tmp"));
            OutputStream fileOutputStream = new FileOutputStream(tmpFile);
            IOUtils.copy(bomInputStream, fileOutputStream);
            fileOutputStream.close();
        }
        bomInputStream.close();

        if (tmpFile != null) {
            candidateXmlfile.delete();
            tmpFile.renameTo(candidateXmlfile);
        }
    }

    public void checkFileCanBeUsed(File candidateXmlFile) throws Exception {
        if (!checkFileMimeTypeRecognised(candidateXmlFile)) {
            logger.warn(
                    MessageFormat.format("N: {0}: {1}", XmlFileFinder.ERROR_FILE_MIMETYPE, candidateXmlFile.getPath()));
            throw new FinderException(XmlFileFinder.ERROR_FILE_MIMETYPE);
        }

        if (checkFileContentsWellFormed(candidateXmlFile) == Boolean.FALSE) {
            logger.warn(MessageFormat.format("N: {0}: {1}", XmlFileFinder.ERROR_FILE_CONTENTS_NOT_WELL_FORMED,
                    candidateXmlFile.getPath()));
            throw new XmlFileFinder.FinderException(XmlFileFinder.ERROR_FILE_CONTENTS_NOT_WELL_FORMED);
        }

        rmBomFromFileContents(candidateXmlFile);
        logger.debug(MessageFormat.format("Y: {0}", candidateXmlFile.getPath()));
    }

    private Boolean checkFileContentsWellFormed(File candidateXmlFile) {
        try {
            DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
            documentBuilderFactory.setValidating(false);
            documentBuilderFactory.setFeature("http://apache.org/xml/features/nonvalidating/load-dtd-grammar", false);
            documentBuilderFactory.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);

            DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();

            documentBuilder.parse(candidateXmlFile);
        } catch (Exception exception) {
            logger.debug(exception.getMessage());
            return false;
        }
        return true;
    }

    private Boolean checkFileMimeTypeRecognised(File candidateXmlFile) throws IOException {
        String mimeType = Files.probeContentType(Paths.get(candidateXmlFile.getPath()));
        logger.debug(mimeType);

        return (mimeType.equals("application/xml") || mimeType.equals("text/xml"));
    }

    public String contentsOfFile(File xmlFile) throws Exception {
        return FileUtils.readFileToString(xmlFile, StandardCharsets.UTF_8);
    }

    @SuppressWarnings("serial")
    public class FinderException extends Exception {
        FinderException(String message) {
            super(message);
        }
    }
}
